Closes #2267 AC WebExtensions + WebCompat + FxAWebChannels

keianhzo opened the pull request:
Closes #2267 Closes #3339 This PR adds support for the following:
- AC 44.0.0
- AC Web Extensions bridge. We are using a `ComponentsAdapter` as a bridge between our `SessionStore` and AC's `BrowserStore` which is required by the AC WebExtensions/Features lifecycle.
- Moves our current builtin extensions (YouTube and Vimeo) to AC `WebExtensionsRuntime`.
- Adds support for the `WebCompat` extension.
- The `SessionStore` is now the `WebExtensionDelegate`
- Session life cycle code refactor to establish the `SessionStore` as the single point of communications with the `BrowserStore`.

There are some Glean related updates, mainly initialization changes and lint errors in .md files, could you please @daoshengmu verify that everything is working as expected?

@bluemarvin @MortimerGoro I've have replaced the appcompat aar for the maven version so it stays aligned with the rest of the androidx components. I couldn't find any speech related classes in that module.

**QA**: 
This should affect mainly to the WebExtensions. Things that we should keep an eye on:
- Youtube and Vimeo WebExtensions lifecycle (correct load of the extensions in restored session)
- FxA login: This shouldn't change, we are now using WebChannels but the flow is still the same.
- Would be also worth keeping an eye on sessions state: creation, recreation, shutdown and restore.

keianhzo commented:
@pocmo As you were involved with the `ComponentsAdapter` implementation, could you please help review? Thanks!

daoshengmu commented:
Please rebase your patch with master first. I was helping you to rebase it in local, but you were working on your own remote, so I can't help you give a commit.

After my investigation, your patch will skip calling `VRBrowserApplication::onActivityCreate()` when running unit tests. That actually breaks our unit test mechanism. I need do more investigation to understand why.

daoshengmu commented:
If we move back
```
        TelemetryWrapper.init(this);
         GleanMetricsService.init(this);
```
to `onCreate()`, the unit tests can be executed. In addition, `Configuration` in Glean seems to be unable to access directly when using Java.

daoshengmu commented:
Btw, we are better not to use the default client uploader. We should use GeckoView client uploader instead. I can work on it at another PR.

https://github.com/MozillaReality/FirefoxReality/issues/3339

keianhzo commented:
@daoshengmu I've fixed the rebase issues.

> BTW, we are better not to use the default client uploader. We should use GeckoView client uploader instead. I can work on it at another PR.

This PR already uses the `GeckoWebExecutor` in the Configuration but it seems that you reverted that in your branch.

The main difference between this PR and master is that it uses the `GeckoWebExecutor` in the Glean configuration which forces a Runtime initialization in the Application's `onCreate` while before we weren't creating the runtime until the Activity's `onCreate`. 

That seems to break Gecko somehow as nothing is renderer anymore and it also breaks the tests with:
> java.lang.IllegalArgumentException: Crash handler service must run in a separate process

As it seems that the `getSystemService` call that happens in the `GeckoRuntime` initialization is not returning the correct process for the Crash handler service. I guess the testing context is not loading the correct manifest information when using the `Application` context?

The weird thing is that the GeckoRuntime docs states:
`Create a new runtime with default settings and attach it to the given context. Create will throw if there is already an active Gecko instance running, to prevent that, bind the runtime to the process lifetime instead of the activity lifetime.` So it should work.

pocmo commented on app/src/common/shared/org/mozilla/vrbrowser/browser/components/GeckoEngineSession.kt:
```diff
@@ -0,0 +1,40 @@
+        val geckoSession: GeckoSession
+): EngineSession() {
+    constructor() :
+        this(SessionStore.get().createSession(false, false).geckoSession)
```
Is this empty constructor, that explicitly creates a session, needed? Looks like that would cause problems because we never want to create one and always use the one you already created in FxR code?

keianhzo replied:
It seems that in some cases the an extension needs to create a new tab and uses that constructor to do so?

https://github.com/mozilla-mobile/android-components/blob/d642612386e8d9880ff83a78737e065b8ac870c7/components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt#L80-L82

        override fun onToggleActionPopup(extension: WebExtension, action: Action): EngineSession? {
            return webExtensionDelegate?.onToggleActionPopup(extension, GeckoEngineSession(runtime), action)
        }

pocmo replied:
Oh, I see! Then I guess this makes sense if that will then let you know about this tab too. :)

keianhzo replied:
Yes, calling `SessionStore.get().createSession` should both add the session to the SessionStore sessions list and to the BrowserStore.

pocmo commented on app/src/common/shared/org/mozilla/vrbrowser/browser/adapter/ComponentsAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+package org.mozilla.vrbrowser.browser.adapter
```
nit: Do you use license headers? I probably didn't add any in my prototype.

pocmo commented on app/src/common/shared/org/mozilla/vrbrowser/browser/components/GeckoEngineSession.kt:
```diff
@@ -0,0 +1,40 @@
+package org.mozilla.vrbrowser.browser.components
```
nit: license?

pocmo commented on app/src/common/shared/org/mozilla/vrbrowser/browser/adapter/ComponentsAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+    companion object {
+        private val instance: ComponentsAdapter = ComponentsAdapter()
+
+        fun get(): ComponentsAdapter = instance
```
nit: You could consider using `@JvmStatic` here so that you can use `ComponentsAdapter.get()` directly from Java.

keianhzo commented:
@daoshengmu I've fixed the GV issue but I still can see issues with test, especially thee issue with the crashservice process not being correctly determined from the manifest in test. I'll have look into that but if you have any ideas, they are welcom.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/components/GeckoWebExtensionRuntime.kt:
```diff
@@ -0,0 +1,281 @@
+                GeckoWebExtension(extension, runtime)
+            } ?: emptyList()
+
+            val privateBrowsingMigrated = sharedPref.getBoolean(MOZAC_HAS_RESET_WEBEXT_PRIVATE_BROWSING_DEFAULT, false)
+            extensions.forEach { extension ->
+                // As a workaround for https://bugzilla.mozilla.org/show_bug.cgi?id=1621385,
+                // we set all installed extensions to be allowed in private browsing mode.
+                // We need to revert back to false which is now the default.
+                if (!extension.isBuiltIn() && extension.isAllowedInPrivateBrowsing() && !privateBrowsingMigrated) {
+                    setAllowedInPrivateBrowsing(extension, false)
+                }
+
```
Do we need this? Looks like some compat code that won't affect us.

keianhzo replied:
Not really. Removed.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/engine/Session.java:
```diff
@@ -521,19 +517,24 @@ private GeckoSession createGeckoSession(@NonNull SessionSettings aSettings) {
         mState = mState.recreate();
 
-        restore();
+        mSessionChangeListeners.forEach(listener -> {
+            listener.onSessionClosed(this);
+            listener.onSessionRemoved(this);
+        });
```
Why are we closing the session and removing it after it is recreated?

keianhzo replied:
Yep, good catch, we don't need to do that, just close the old one and open the new one. I've updated.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/engine/SessionStore.java:
```diff
@@ -89,15 +112,29 @@ public void onTrackingProtectionLevelUpdated(int level) {
         mHistoryStore = new HistoryStore(context);
+
+        // Web Extensions initialization
+        YoutubeExtension.install(mWebExtensionRuntime);
+        VimeoExtension.install(mWebExtensionRuntime);
```
This pattern doesn't scale. If the WebExtenstionRuntime API changes we have to update every class. I especially dislike it because the only difference between the two classes is the meta data.

keianhzo replied:
We can add a builtin extensions manager that initializes all of them based on metadata as we had before. I don't expect that we have many builtin extensions though mot of them are going to come from AC or the Addon Manager in the future, so I'm not sure if it'll be that useful.

bluemarvin replied:
Or just have a single class that takes the meta data in the constructor. I guess it doesn't matter.

keianhzo replied:
Fixed, I've added a BuiltinExtension class that handles the initialization for our builtin extensions.

bluemarvin commented on app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java:
```diff
@@ -738,5 +736,15 @@ public void setDownloadsSortingOrder(@SortingContextMenuWidget.Order int order)
     public @Storage int getDownloadsSortingOrder() {
         return mPrefs.getInt(mContext.getString(R.string.settings_key_downloads_sorting_order), DOWNLOADS_SORTING_ORDER_DEFAULT);
     }
+
+    public void setFxAWebChannelsEnabled(boolean isEnabled) {
+        SharedPreferences.Editor editor = mPrefs.edit();
+        editor.putBoolean(mContext.getString(R.string.settings_key_webchannels_enabled), isEnabled);
+        editor.commit();
+    }
+
+    public boolean isFxAWebChannelsEnabled() {
+        return mPrefs.getBoolean(mContext.getString(R.string.settings_key_webchannels_enabled), FXA_WEBCHANNELS_ENABLED);
+    }
```
Is there a reason we want to be able to turn WebChannels off? For QA and testing?

keianhzo replied:
I just left it just in case we want to enable/disable easily in the future. I can remove it, is not really used right now.

bluemarvin replied:
I wasn't sure if there was a reason. I would remove it if we don't want to support the previous version.

keianhzo replied:
Removed

daoshengmu commented:
> @daoshengmu I've fixed the GV issue but I still can see issues with test, especially thee issue with the crashservice process not being correctly determined from the manifest in test. I'll have look into that but if you have any ideas, they are welcomed.

You can run `./gradlew app:testNoapiArm64DebugUnitTest` to execute the test. Or just pick one `testXXX ` from `./gradlew tasks`.

So far, I notice there is no one instances `VRBrowserActivity`, but it do enter `VRBrowserApplication::onCreate()`.

keianhzo commented:
@daoshengmu I have fixed the Telemetry tests. The GV team recommended to move the GeckoRuntime startup code to the Activity onCreate so I moved all the initialization there and initialized the Telemetry services for tests in a setup method. Seems to be working fine but I needed to add a flag to avoid the legacy LegacyTelemetry initialization.

daoshengmu commented on app/src/common/shared/org/mozilla/vrbrowser/VRBrowserApplication.java:
```diff
@@ -50,6 +47,16 @@ protected void onActivityCreate() {
+        );
+        GleanMetricsService.init(this,
+                EngineProvider.INSTANCE.getDefaultClient(this),
+                true
```
Remove `true` flag

daoshengmu commented on app/src/common/shared/org/mozilla/vrbrowser/browser/Services.kt:
```diff
@@ -112,6 +120,7 @@ class Services(val context: Context, places: Places): GeckoSession.NavigationDel
 
     ).also {
         it.registerForAccountEvents(deviceEventObserver, ProcessLifecycleOwner.get(), true)
+
```
Redundant newline?

daoshengmu commented on app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java:
```diff
@@ -68,8 +72,14 @@ public static void init(Context aContext) {
 
-        LegacyTelemetry.INSTANCE.clientId().set(UUID.fromString(TelemetryHolder.get().getClientId()));
-        Configuration config = new Configuration(Configuration.DEFAULT_TELEMETRY_ENDPOINT, BuildConfig.BUILD_TYPE);
+        if (initializeLegacy) {
+            LegacyTelemetry.INSTANCE.clientId().set(UUID.fromString(TelemetryHolder.get().getClientId()));
+        }
```
```suggestion
        LegacyTelemetry.INSTANCE.clientId().set(UUID.fromString(TelemetryHolder.get().getClientId()));
```

keianhzo commented on app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java:
```diff
@@ -54,7 +58,7 @@
 
     // We should call this at the application initial stage.
-    public static void init(Context aContext) {
+    public static void init(@NonNull Context aContext, @NonNull Client client, boolean initializeLegacy) {
```
ok, I wasn't sure if that test was important. I've removed it and got rid of the legacy initialization flag.

keianhzo commented on app/src/common/shared/org/mozilla/vrbrowser/browser/engine/SessionStore.java:
```diff
@@ -359,4 +443,139 @@ public void removePermissionException(@NonNull String uri, @SitePermission.Categ
+    @Override
+    public void onAllowedInPrivateBrowsingChanged(@NonNull WebExtension webExtension) {
+
+    }
+
+    @Override
+    public void onNewTab(@NonNull WebExtension webExtension, @NonNull EngineSession engineSession, boolean b, @NonNull String s) {
+
+    }
+
+    @Override
+    public void onBrowserActionDefined(@NonNull WebExtension webExtension, @NonNull Action action) {
+
+    }
+
+    @Override
+    public void onPageActionDefined(@NonNull WebExtension webExtension, @NonNull Action action) {
+
+    }
+
+    @Nullable
+    @Override
+    public EngineSession onToggleActionPopup(@NonNull WebExtension webExtension, @NonNull EngineSession engineSession, @NonNull Action action) {
+        return null;
+    }
+
+    @Override
+    public boolean onInstallPermissionRequest(@NonNull WebExtension webExtension) {
+        return false;
+    }
+
+    @Override
+    public void onUpdatePermissionRequest(@NonNull WebExtension webExtension, @NonNull WebExtension webExtension1, @NonNull List<String> list, @NonNull Function1<? super Boolean, Unit> function1) {
+
+    }
+
+    @Override
+    public void onExtensionListUpdated() {
+
+    }
```
They don't have default implementations for that interface. We don't use them right now but we will when we add support for the addon manager.

keianhzo commented on app/src/common/shared/org/mozilla/vrbrowser/browser/engine/Session.java:
```diff
@@ -454,7 +448,7 @@ private void loadDefaultPage() {
     }
 
-    private void restore() {
+    private void restore(boolean isRecreation) {
```
It seems that the SessionState is null when is not a recreation. I've removed the flag.
